src/role.rs

```diff
@@ -1,88 +1,64 @@
-use serenity::prelude::Context;
+use rand::{seq::IteratorRandom, thread_rng};
+use serenity::{
+    all::{Member, Permissions},
+    builder::EditRole,
+    http::CacheHttp,
+    prelude::Context,
+};
+use std::{collections::HashMap, hash::Hash};
 
 fn rand_hash<K: Eq + Hash, V>(hash: &HashMap<K, V>) -> &K {
-    let key = hash.keys().choose(&mut thread_rng()).unwrap();
-
-    key
+    hash.keys().choose(&mut thread_rng()).unwrap()
 }
 
 pub async fn random_color(ctx: &Context, member: &Member) -> Result<(), Error> {
-    let choice = rand_hash(&CONFIG.colors);
-
     let guild = match member.guild_id.to_guild_cached(&ctx.cache) {
         Some(g) => g.clone(),
-        None => return Ok(())
+        None => return Ok(()),
     };
 
-    let r = EditRole::new().name(choice.as_str()).colour(CONFIG.colors[choice]).permissions(Permissions::empty());
-    let role_id = match guild.role_by_name(choice.as_str()) {
+    let choice = rand_hash(&CONFIG.colors);
+    let r = EditRole::new().name(choice).colour(CONFIG.colors[choice]).permissions(Permissions::empty());
+    let role_id = match guild.role_by_name(choice) {
         Some(role) => role.id,
-        None => {
-            let r = guild.create_role(&ctx.http, r).await.unwrap();
-
-            r.id
-        }
+        None => guild.create_role(&ctx.http, r).await.unwrap().id,
     };
 
-    let m = member.clone();
     match member.roles(&ctx.cache) {
         Some(roles) => {
-            for role in roles {
-                if CONFIG.colors.contains_key(role.name.as_str()) {
-                    m.remove_role(ctx.http(), role.id).await?;
-                    break;
-                }
+            for role in roles.iter().filter(|role| CONFIG.colors.contains_key(&role.name)) {
+                member.remove_role(ctx.http(), role.id).await?
             }
         }
         None => {}
     };
 
-    m.add_role(&ctx.http, role_id).await?;
-    Ok(())
+    Ok(member.add_role(&ctx.http, role_id).await?)
 }
 
-pub async fn process_color(ctx: PoiseContext<'_>, choice: String) -> Result<(), Error> {
+pub async fn process_color(ctx: PoiseContext<'_>, choice: &str) -> Result<(), Error> {
     let guild = match ctx.guild() {
         Some(guild) => guild.clone(),
-        None => {
-            eprintln!("Can't find server ...");
-            return Ok(())
-        }
+        None => return Ok(eprintln!("Can't find server ...")),
     };
 
-    let r = EditRole::new().name(choice.as_str()).colour(CONFIG.colors[&choice]).permissions(Permissions::empty());
-    let role_id = match guild.role_by_name(choice.as_str()) {
+    let r = EditRole::new().name(choice).colour(CONFIG.colors[choice]).permissions(Permissions::empty());
+    let role_id = match guild.role_by_name(choice) {
         Some(role) => role.id,
-        None => {
-            let r = guild.create_role(ctx.http(), r).await?;
-
-            r.id
-        }
+        None => guild.create_role(ctx.http(), r).await?.id,
```

Comment:
as far as I can see, `r` is only needed in the `None` case, so we should move it there